python/lsst/ctrl/mpexec/simple_pipeline_executor.py

```diff
@@ -0,0 +1,303 @@
+
+from typing import Any, Iterable, Iterator, Optional, Type, Union, List
+
+from lsst.daf.butler import Butler, CollectionType, Quantum
+from lsst.obs.base import Instrument
+from lsst.pex.config import Config
+from lsst.pipe.base import GraphBuilder, Pipeline, PipelineTask, QuantumGraph, TaskDef
+
+from .preExecInit import PreExecInit
+from .singleQuantumExecutor import SingleQuantumExecutor
+from .taskFactory import TaskFactory
+
+
+class SimplePipelineExecutor:
+    """A simple, high-level executor for pipelines.
+
+    Parameters
+    ----------
+    quantum_graph : `QuantumGraph`
+        Graph to be executed.
+    butler : `Butler`
+        Object that manages all I/O.  Must be initialized with `collections`
+        and `run` properties that correspond to the input and output
+        collections, which must be consistent with those used to create
+        ``quantum_graph``.
+
+    Notes
+    -----
+    Most callers should use one of the `classmethod` factory functions
+    (`from_pipeline_filename`, `from_task_class`, `from_pipeline`) instead of
+    invoking the constructor directly; these guarantee that the `Butler` and
+    `QuantumGraph` are created consistently.
+
+    This class is intended primarily to support unit testing and small-scale
+    integration testing of `PipelineTask` classes.  It deliberately lacks many
+    features present in the command-line-only ``pipetask`` tool in order to
+    keep the implementation simple.  Python callers that need more
+    sophistication should call lower-level tools like `GraphBuilder`,
+    `PreExecInit`, and `SingleQuantumExecutor` directly.
+    """
+
+    def __init__(self, quantum_graph: QuantumGraph, butler: Butler):
+        self.quantum_graph = quantum_graph
+        self.butler = butler
+
+    @classmethod
+    def prep_butler(
+        cls, root: str, inputs: Iterable[str], output: str, output_run: Optional[str] = None,
+        writeable: Optional[bool] = False
+    ) -> Butler:
+        """Helper method for creating `Butler` instances with collections
+        appropriate for processing.
+
+        Parameters
+        ----------
+        root : `str`
+            Root of the butler data repository; must already exist, with all
+            necessary input data.
+        inputs : `Iterable` [ `str` ]
+            Collections to search for all input datasets, in search order.
+        output : `str`
+            Name of a new output `~CollectionType.CHAINED` collection to create
+            that will combine both inputs and outputs.
+        output_run : `str`, optional
+            Name of the output `~CollectionType.RUN` that will directly hold
+            all output datasets.  If not provided, a name will be created from
+            ``output`` and a timestamp.
+        writeable: `bool`, optional
+            Make the butler a read/write butler; we can't make this automatic,
+            because the output collection(s) don't exist yet.
```

Comment:
I don't understand the comment "we can't make this automatic, because the output collection(s) don't exist yet".  Also, how can this work at all if `writeable=False`?

Reply:
Ah, I think this was a miscommunication between @parejkoj about how he should resolve a bug in my first draft.  The usefulness is extremely narrow if `writeable=False` (you can construct the class and in doing so build a `QuantumGraph`, but then can't use it), so I'm just going to remove this parameter and force `writeable=True` in the `Butler` construction line.